File: app-flutter/lib/core/firebase/firestore_sync_config.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/foundation.dart';

/// تفعيل كاش Firestore على القرص للمنصات الأصلية. الويب لا يدعم persistence هنا،
/// لذلك نعتمد هناك على Hive metadata cache الذي يسبق أي انتظار للشبكة.
class FirestoreSyncConfig {
  const FirestoreSyncConfig._();

  /// قاعدة Firestore المسمّاة في مشروع nebras-9118c (بدون أقواس).
  ///
  /// لوحة التحكم والـ Admin SDK يستخدمان `getFirestore(app, 'default')`.
  /// `FirebaseFirestore.instance` على الموبايل يتّصل افتراضياً بـ `(default)`
  /// (مع أقواس) — قاعدة مختلفة غير موجودة، فيرجع NOT_FOUND وتبقى الشاشة فارغة
  /// بينما يعمل FCM لأنّه لا يمرّ عبر Firestore.
  static const String nebrasDatabaseId = 'default';

  static FirebaseFirestore get instance => FirebaseFirestore.instanceFor(
        app: Firebase.app(),
        databaseId: nebrasDatabaseId,
      );

  static Future<void> configure(FirebaseFirestore firestore) async {
    if (kIsWeb) return;
    try {
      firestore.settings = const Settings(
        persistenceEnabled: true,
        cacheSizeBytes: 100 * 1024 * 1024,
      );
    } catch (e) {
      debugPrint('[FirestoreSyncConfig] persistence already locked: $e');
    }
  }

  /// قراءة تستفيد من الكاش ثم تتزامن مع الخادم (`serverAndCache`) حتى لا تبقى
  /// وثائق محذوفة في الواجهة عند عودة الشبكة.
  static Future<QuerySnapshot<Map<String, dynamic>>> cacheFirstFresh(
    Query<Map<String, dynamic>> query,
  ) =>
      query.get(const GetOptions(source: Source.serverAndCache));
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/firebase/FirestoreSyncConfig.kt
package com.nebras.mobile.core.firebase

import android.util.Log
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.google.firebase.firestore.PersistentCacheSettings
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.Source
import kotlinx.coroutines.tasks.await

/**
 * تفعيل كاش Firestore على القرص.
 *
 * ⚠️ **قاعدة Firestore المسمّاة**: مشروع `nebras-9118c` يستعمل قاعدة اسمها
 * `default` (بدون أقواس). اللوحة والـ Admin SDK يستخدمان
 * `getFirestore(app, 'default')`. النسخة الافتراضية على الجوّال تتّصل بـ
 * `(default)` (مع أقواس) — قاعدة مختلفة غير موجودة، فيرجع NOT_FOUND وتبقى
 * الشاشة فارغة بينما يعمل FCM لأنّه لا يمرّ عبر Firestore.
 */
object FirestoreSyncConfig {

    private const val TAG = "FirestoreSyncConfig"

    /** قاعدة Firestore المسمّاة في مشروع nebras-9118c (بدون أقواس). */
    const val NEBRAS_DATABASE_ID = "default"

    private const val CACHE_SIZE_BYTES = 100L * 1024 * 1024

    val instance: FirebaseFirestore by lazy {
        FirebaseFirestore.getInstance(FirebaseApp.getInstance(), NEBRAS_DATABASE_ID)
            .also(::configure)
    }

    /** يجب استدعاؤه قبل أوّل قراءة — الإعدادات تُقفل بعد أول استعمال. */
    fun configure(firestore: FirebaseFirestore) {
        runCatching {
            firestore.firestoreSettings = FirebaseFirestoreSettings.Builder()
                .setLocalCacheSettings(
                    PersistentCacheSettings.newBuilder()
                        .setSizeBytes(CACHE_SIZE_BYTES)
                        .build(),
                )
                .build()
        }.onFailure {
            Log.d(TAG, "persistence already locked: ${it.message}")
        }
    }

    /**
     * قراءة تستفيد من الكاش ثم تتزامن مع الخادم حتى لا تبقى وثائق محذوفة
     * في الواجهة عند عودة الشبكة (مقابل `Source.serverAndCache` في Dart).
     */
    suspend fun cacheFirstFresh(query: Query): QuerySnapshot =
        query.get(Source.DEFAULT).await()
}
